[#6187][feat] add LayerNorm module

Funatiq opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a customizable Layer Normalization module with support for optional residual connections and configurable parameters.

* **Style**
  * Improved code readability and formatting in the RMSNorm module.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up
-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Funatiq committed:
[style] Refactor RMSNorm constructor for improved readability by adjusting parameter formatting

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[feat] Add LayerNorm module for improved normalization in Torch models

- Introduce a new LayerNorm class, implementing layer normalization with optional weights and biases.
- The forward method handles input tensors and supports residual connections, while the skip_forward method allows bypassing normalization.
- The implementation includes checks for Flashinfer availability.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
Add copyright header

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[docs] Add missing docstrings for the LayerNorm class and methods.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented on tensorrt_llm/_torch/modules/layer_norm.py:
```diff
@@ -0,0 +1,125 @@
+            hidden_states = (hidden_states - mean
+                             ) * torch.rsqrt(variance + self.variance_epsilon)
+            hidden_states = self.weight * hidden_states.to(
+                input_dtype) + self.bias
```
Copied the type conversion from `RMSNorm` but this seems strange to me. Should it conversion happen for the final result?

yuxianq commented on tensorrt_llm/_torch/modules/layer_norm.py:
```diff
@@ -0,0 +1,118 @@
+        else:
+            return hidden_states, residual
+
+    def skip_forward(
```
It is used to skip modules in pp cases. When `pp_size > 1`, only the last rank runs the last norm layer, so we replace the `forward` function with `skip_forward` function for the last norm layer in other ranks to reduce redundant computation.

Funatiq commented:
/bot run

bobboli commented on tensorrt_llm/_torch/modules/layer_norm.py:
```diff
@@ -0,0 +1,118 @@
+            hidden_states = hidden_states + residual.to(torch.float32)
+            residual = hidden_states.to(input_dtype)
+
+        mean = hidden_states.mean(-1, keepdim=True)
```
This is an unfused vanilla implementation. Unfortunately flashinfer only provides RMSNorm, no LayerNorm.
Just curious, why not directly use [torch.nn.LayerNorm](https://docs.pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html)? I am not sure whether there is an optimized kernel underneath the module though.

Funatiq replied:
So if I understand correctly, we don't actually need to add our own `LayerNorm` module and we can simply use the native torch.nn.LayerNorm. Seems like that's what we are doing anyway in the model definitions.

bobboli replied:
You mean this [else branch](https://github.com/NVIDIA/TensorRT-LLM/blob/4f8afe4cc6ef9a8ec3763852a6c0f8e5f9800a51/tensorrt_llm/_torch/modules/rms_norm.py#L45 )? Yeah I think it is fine to replace it with torch. AFAIK that else branch is seldom used, so I think it doesn't matter much.

Funatiq committed:
refactor: simplify LayerNorm forward method using nn.functional.layer_norm

- Replaced manual mean and variance calculations with nn.functional.layer_norm for improved readability and performance.
- Maintained support for weight, bias, and variance epsilon parameters.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #14959](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14959/) [ run ] triggered by Bot

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #14975](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14975/) [ run ] triggered by Bot
